### PR TITLE
Solving PNP works

### PR DESCRIPTION
**Pull request by arthbous**
@maximilianmetti 

To test it just:
`python pnp.py`

### CODE REVIEW

File: pnp.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+CheckDir(DATA_DIR, CLEAN)
+CheckDir(IMG_DIR, CLEAN)
+
+# Create mesh and define function space
+Lx = 10.0
+Ly = 2.0
+Lz = 2.0
+P1 = Point(-Lx/2.0, -Ly/2.0, -Lz/2.0)
+P2 = Point(Lx/2.0, Ly/2.0, Lz/2.0)
+mesh = BoxMesh(P1, P2, 50, 5, 5)
```
Comment (maximilianmetti) [inline pnp.py]: This box is a `20*4*4`, and there are `50*5*5` dofs distributed. Did you mean to make the mesh nonuniform?  A more uniform mesh would be `50*10*10` or `25*5*5`.

File: LinearExp.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+#! /usr/bin/python2.7
+"""
+Linear Expression Class.
+    You have two choices: the python version or the C++
+    Python:
+        Exp = Linear_Function(0, -Lx/2.0, Lx/2.0, 0.0, -2.0, degree=2)
+    C++:
+        Exp = Expression(LinearFunction_cpp, degree=2)
+        CationExpression .update(0, -Lx/2.0, Lx/2.0, 0.0, -2.0)
```
Comment (maximilianmetti) [inline LinearExp.py:9]: - This isn't clear to me what the difference of the C++ and python invocations are; is one faster?
 - How is the `degree=2` used? is it tolerated as an extra parameter (due to `kwargs`) and simply discarded?

File: pnp.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    - (eps * inner(grad(EsEs), grad(phi))) * dx \
+    - (- (qp * exp(CatCat)*CatCat + qn * exp(AnAn) * AnAn) * phi)*dx
+
+# Parameters
+tol = 1E-8
+itmax = 20
+it = 0
+u0 = Constant((0.0, 0.0, 0.0))
+bc = DirichletBC(V, u0, boundary)
+solver = PETScKrylovSolver("gmres", "ilu")
```
Comment (maximilianmetti) [inline pnp.py]: This does not work if `parameters["linear_algebra_backend"] = "Eigen"` is used.  There should be some logic around this statement to ensure the solver does not crash. I.e. I ran this using `Eigen` as `Petsc` was not available with my version of dolfin and this line threw an error.

**Review by maximilianmetti — COMMENTED** [review]
This is a good start; we need to fix the error regarding the linear solver and should certainly include more information in the readme regarding the setup of this repo.  

The first thing we need to have is a version management systme.  This will help ensure the `dolfin` version is correct (along with any other packages); I have to update my dolfin version to get this working.

Is there any reason why we're using an old version of python? There are quite a few breaking changes between python 2 and 3. I think using python 3 will lead to better support in the future, no?

Let's chat about this tomorrow and see if we can get this up and running on my machine (and document the process).

File: LinearExp.py (hunk 2 above)
Comment (arthbous) [reply LinearExp.py:9]: With the new version of fenics you have to specify the degree of interpolation now. The invocations in C++ are faster so we should keep that one.

File: pnp.py (hunk 1 above)
Comment (arthbous) [reply pnp.py]: Yes, I am changing it.

File: pnp.py (hunk 3 above)
Comment (arthbous) [reply pnp.py]: that's because i am using PETSc solver, I will see if I can define a general solver.

**Comment by arthbous** [comment]
As for Python 2 or 3. I prefer 2 by far it has more libraries.

**Comment by arthbous** [comment]
@maximilianmetti I made all the changes

**Pull request merged**